### PR TITLE
[fox] Initial contribution

### PR DESCRIPTION
**Pull request by ksubzda-nxw**
[fox] Initial contribution including basic integration with Fox Systems automation

This is an initial contribution of Fox Binding. The binding integrates with Fox System by Nexwell Engineering which is a smart home solution composed of devices controlling lights, rollers, and general electric outputs, as well as reacting on wall switches and sensors. The system has its own inner programmable automation via FoxMaker application (available with full functionality for free on http://nexwell.eu/fox-software/foxmaker/). In the application, user can create a set of so-called Tasks and Results, which can further constitute an API of the particular system. This API - in the form of JSON description - can be added to a Fox thing via Fox Binding in its configuration field. Then Fox thing is created with dynamically loaded commands and triggers, based on - respectively - tasks and results in the system itself. This solution creates now therefore a two-way integration between openHAB and Fox, at the level of their automation.

The source code is contributed and will be further maintained officially by the company, and is meant to be and stay open source. Future development will possibly include addition of Fox System devices as things, and their direct control by openHAB.

### CODE REVIEW

**Comment by ksubzda-nxw** [comment]
> It will be awesome if you fix what is wrong and commit the changes.

Is there an option in Eclipse configuration of the project, that would display these errors? They just seem to follow some more restricted rules of coding that I was not aware of, and Eclipse did not even mention if there is even any problem like that (I have no errors and no warnings in the project).

**Comment by ksubzda-nxw** [comment]
I should probably use https://github.com/openhab/static-code-analysis to check for these errors. Am i right?

**Comment by fwolter** [comment]
Could you fix the build errors? I would take a look into it, then.

**Comment by ksubzda-nxw** [comment]
> 
> 
> Could you fix the build errors? I would take a look into it, then.

Build works for me with no errors. So I don't know what errors you are referring to.

File: bundles/org.openhab.binding.fox/src/main/java/org/openhab/binding/fox/internal/core/FoxMessage.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        return String.format("@%s:%s %s", devToken, appToken, message);
+    }
+
+    void interpret(String data) {
+        reset();
+        if (data.matches("@[^ ]+:[^ ]+ .+")) {
+            appToken = data.substring(data.indexOf("@") + 1, data.indexOf(":"));
+            devToken = data.substring(data.indexOf(":") + 1, data.indexOf(" "));
+            message = data.substring(data.indexOf(" ") + 1);
+        }
```
Comment (fwolter) [inline bundles/org.openhab.binding.fox/src/main/java/org/openhab/binding/fox/internal/core/FoxMessage.java:65]: If you already use regex, why not extract the values you want with a regex pattern matcher? See https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html#matcher(java.lang.CharSequence)

File: bundles/org.openhab.binding.fox/src/main/java/org/openhab/binding/fox/internal/core/FoxMessengerTcpIp.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+ * The {@link FoxMessengerTcpIp} implements communication with Fox system via TCP/IP telnet service.
+ *
+ * @author Kamil Subzda - Initial contribution
+ */
+@NonNullByDefault
+public class FoxMessengerTcpIp implements FoxMessenger {
+
+    private String host = "";
+    private int port = 0;
+    private int timeout = 250;
```
Comment (fwolter) [inline bundles/org.openhab.binding.fox/src/main/java/org/openhab/binding/fox/internal/core/FoxMessengerTcpIp.java:35]: It's good practice to append the unit to the field name e.g. timeoutMs

File: bundles/org.openhab.binding.fox/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		<label>Fox</label>
+		<description>Nexwell Fox System</description>
+		<channels>
+			<channel id="taskCommand" typeId="taskRequest" />
+			<channel id="taskState" typeId="taskExecuted" />
+			<channel id="resultTrigger" typeId="resultNotice" />
+			<channel id="resultState" typeId="resultAcquired" />
+		</channels>
+		<config-description>
+			<parameter name="gateHost" type="text" required="true">
```
Comment (fwolter) [inline bundles/org.openhab.binding.fox/src/main/resources/ESH-INF/thing/thing-types.xml:17]: You could specify `<context>network-address</context>` to get a validation in PaperUI.

File: bundles/org.openhab.binding.fox/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+				<label>Connection address</label>
```
Comment (fwolter) [inline bundles/org.openhab.binding.fox/src/main/resources/ESH-INF/thing/thing-types.xml:18]: Words in labels should be capitalized. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

File: bundles/org.openhab.binding.fox/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+				<description>Fox System net address - IP or domain, with an optional port (default: 23).</description>
```
Comment (fwolter) [inline bundles/org.openhab.binding.fox/src/main/resources/ESH-INF/thing/thing-types.xml:19]: Better use a dedicated config parameter for the port and apply the default value of 23 to it. That also simplifys code.

File: bundles/org.openhab.binding.fox/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+			</parameter>
+			<parameter name="gatePassword" type="text" required="false">
+				<label>Connection password</label>
```
Comment (fwolter) [inline bundles/org.openhab.binding.fox/src/main/resources/ESH-INF/thing/thing-types.xml:22]: Add `<context>password</context>`.

File: bundles/org.openhab.binding.fox/README.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# Fox Binding
+
+This binding integrates with Nexwell Fox Systems. It allows to request system Tasks executions, and to notice system Results returns. The binding connects to the system via its telnet service.
+
+## Supported Things
+
+As for now one thing is supported, which is Fox, representing Fox System as a whole.
```
Comment (fwolter) [inline bundles/org.openhab.binding.fox/README.md:7]: Can you mention the ThingID `system` here?

**Comment by kaikreuzer** [comment]
Closing this PR as it is not in a mergeable state. Please follow [the instructions to port this PR to the main branch](https://github.com/openhab/openhab-addons/issues/8512) for openHAB 3.

**Pull request closed without merging**